File: rtbrpgsca/src/costume.rs
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/
use crate::role::*;
use crate::actor::*;
use crate::ability::*;
use crate::state::*;
use crate::scene::*;
use crate::play::*;

use std::collections::HashMap;

#[derive(Clone)]
pub struct Costume<'a> {
    pub(crate) base: Box<Role<'a>>,
    pub(crate) a_skills: Option<Vec<&'a Ability<'a>>>,
    pub(crate) counters: Option<Vec<&'a Ability<'a>>>,
    pub(crate) st_res: Option<HashMap<&'a State<'a>, i32>>,
    pub(crate) res: Option<HashMap<i32, i32>>,
    pub(crate) m_actions: i32,
    pub(crate) cvr_type: i32,
    pub(crate) rfl_type: i32,
    pub(crate) atk: i32,
    pub(crate) def: i32,
    pub(crate) spi: i32,
    pub(crate) wis: i32,
    pub(crate) agi: i32,
}

extend_struct!(Costume, Role);

impl<'a> Costume<'a> {
    
    pub const CAUSES_TXT: &'static str = ", {0} is affected by {1}";

    pub const FLAG_ENRAGED: i32 = 4;
    pub const FLAG_CONFUSE: i32 = 8;
    pub const FLAG_CONVERT: i32 = 16;
    pub const FLAG_SHAPE_SHIFT: i32 = 32;
    pub const FLAG_INVINCIBLE: i32 = 64;
    pub const FLAG_DRAW: i32 = 128;
    pub const FLAG_STUN: i32 = 256;
    pub const FLAG_KO: i32 = 512;

    #[inline(always)]
    pub fn enraged(&self) -> bool {
        self.has_flag(Costume::FLAG_ENRAGED)
    }

    #[inline(always)]
    pub fn confused(&self) -> bool {
        self.has_flag(Costume::FLAG_CONFUSE)
    }

    #[inline(always)]
    pub fn shape_shifted(&self) -> bool {
        self.has_flag(Costume::FLAG_SHAPE_SHIFT)
    }

    #[inline(always)]
    pub fn invincible(&self) -> bool {
        self.has_flag(Costume::FLAG_INVINCIBLE)
    }

    #[inline(always)]
    pub fn stunned(&self) -> bool {
        self.has_flag(Costume::FLAG_STUN)
    }

    #[inline(always)]
    pub fn drawn(&self) -> bool {
        self.has_flag(Costume::FLAG_DRAW)
    }

    #[inline(always)]
    pub fn knocked_out(&self) -> bool {
        self.has_flag(Costume::FLAG_KO)
    }

    #[inline(always)]
    pub fn rfl_type(&self) -> i32 {
        self.rfl_type
    }

    #[inline(always)]
    pub fn cvr_type(&self) -> i32 {
        self.cvr_type
    }

    #[inline(always)]
    pub fn atk(&self) -> i32 {
        self.atk
    }

    #[inline(always)]
    pub fn def(&self) -> i32 {
        self.def
    }

    #[inline(always)]
    pub fn wis(&self) -> i32 {
        self.wis
    }

    #[inline(always)]
    pub fn spi(&self) -> i32 {
        self.spi
    }

    #[inline(always)]
    pub fn agi(&self) -> i32 {
        self.agi
    }

    #[inline(always)]
    pub fn m_actions(&self) -> i32 {
        self.m_actions
    }

    #[inline(always)]
    pub fn a_skills(&self) -> Option<&'a Vec<&'a Ability>> {
        self.a_skills.as_ref().clone()
    }

    #[inline(always)]
    pub fn a_skills_mut(&mut self) -> Option<&'a mut Vec<&'a Ability>> {
        if let Some(v) = self.a_skills.as_mut() {
            Some(v)
        } else {
            None
        }
    }

    #[inline(always)]
    pub fn counters(&self) -> Option<&'a Vec<&'a Ability>> {
        self.counters.as_ref().clone()
    }

    #[inline(always)]
    pub fn counters_mut(&mut self) -> Option<&'a mut Vec<&'a Ability>> {
        if let Some(v) = self.counters.as_mut() {
            Some(v)
        } else {
            None
        }
    }

    #[inline(always)]
    pub fn res(&self) -> Option<&HashMap<i32, i32>> {
        self.res.as_ref().clone()
    }

    #[inline(always)]
    pub fn res_mut(&mut self) -> Option<&mut HashMap<i32, i32>> {
        if let Some(v) = self.res.as_mut() {
            Some(v)
        } else {
            None
        }
    }

    #[inline(always)]
    pub fn st_res(&self) -> Option<&HashMap<&'a State<'a>, i32>> {
        self.st_res.as_ref().clone()
    }

    #[inline(always)]
    pub fn st_res_mut(&mut self) -> Option<&mut HashMap<&'a State<'a>, i32>> {
        if let Some(v) = self.st_res.as_mut() {
            Some(v)
        } else {
            None
        }
    }

    pub(crate) fn refresh<'b>(&self, ret: &mut Option<&mut String>, scene: &mut Option<&mut dyn Scene>, actor: &'b mut Actor, upd_states: bool, remove: bool) {

    }

    pub(crate) fn adopt<'b>(&self, ret: &mut Option<&mut String>, scene: &mut Option<&mut dyn Scene>, actor: &'b mut Actor, upd_states: bool, remove: bool) {

    }

    pub(crate) fn apply(&self, ret: &mut Option<&mut String>, scene: &Option<&mut dyn Scene>, actor: &mut Actor) {
        
    }
    
}

File: qtbrpgsca/costume.cpp
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/
#include "costume.h"
#include "ability.h"
#include "actor.h"
#include "ailment.h"
#include "scene.h"
#include "role.h"

#include <QStringBuilder>

using namespace tbrpgsca;

QString Costume::CausesTxt = ", %1 is affected by %2";

Ability Costume::addedSkill(int const n) const
{
    return (this->costumeData()._a_skills->at(n));
}

bool Costume::hasAddedSkill(const Ability& skill) const
{
    auto aSkills = this->costumeData()._a_skills;
    return aSkills && aSkills->contains(skill);
}

int Costume::addedSkillsSize() const
{
    auto aSkills = this->costumeData()._a_skills;
    return aSkills == NIL ? 0 : aSkills->size();
}

Ability Costume::counterSkill(int const n) const
{
    return (this->costumeData()._counters->at(n));
}

bool Costume::hasCounterSkill(const Ability& skill) const
{
    auto counters = this->costumeData()._counters;
    return counters && counters->contains(skill);
}

int Costume::counterSkillsSize() const
{
    auto counters = this->costumeData()._counters;
    return counters == NIL ? 0 : counters->size();
}

int Costume::elementResistance(const int element) const
{
    QMap<int, int>* res = this->costumeData()._res;
    return res == NIL ? 0 : res->value(element, 0);
}

int Costume::stateResistance(const int state) const
{
    auto stRes = this->costumeData()._st_res;
    return stRes == NIL ? 0 : stRes->value(state, 0);
}

bool Costume::isCountering() const
{
    auto& counters = this->costumeData()._counters;
    return counters && counters->size() > 0;
}

void Costume::adopt(QString& ret, Actor& actor) const
{
    this->adopt(&ret, actor, true, false);
}

void Costume::abandon(QString& ret, Actor& actor) const
{
    this->adopt(&ret, actor, true, true);
}

void Costume::adopt(QString* const ret, Actor& actor, bool const updStates, bool const remove) const
{
    const Costume& costume = *this;
    auto& actorData = actor.actorMutData();
    auto& costumeData = costume.costumeData();
    actor.updateAttributes(remove, costume);
    actor.updateResistance(remove, costumeData._res, costumeData._st_res);
    if (remove)
    {
        int const roleFlags = costume.playFlags();
        int const actorFlags = actor.playFlags();
        if ((actorFlags & roleFlags) == roleFlags)
        {
            actor.setPlayFlags(actorFlags ^ roleFlags);
        }
        if (costume.isShapeShifted() && costume.roleData()._sprite)
        {
            (*actor.roleMutData()._sprite) = *(actor.job().roleMutData()._sprite);
        }
        if (costume.currentHp() == 0 && costume.currentMp() == 0 && costume.currentRp() == 0)
        {
            int const roleElm = costume.damageType();
            int const actorElm = actor.damageType();
            if ((actorElm & roleElm) == roleElm)
            {
                actor.setDamageType(actorElm ^ roleElm);
            }
        }
        else
        {
            QVector<const Costume*>* const dmgRoles = actorData._dmg_roles;
            if (dmgRoles)
            {
                dmgRoles->removeOne(&costume);
            }
        }
        costume.refresh(ret, actor, updStates, true);
        actor.refreshCostumes(ret);
    }
    else
    {
        if (updStates)
        {
            QList<Ailment>* const cStates = costume.roleData()._a_states;
            if (cStates)
            {
                actor.updateStates(false, ret, *cStates, true);
            }
        }
        if (costume.currentHp() != 0 || costume.currentMp() != 0 || costume.currentRp() != 0)
        {
            QVector<const Costume*>* dmgRoles = actorData._dmg_roles;
            if (dmgRoles == NIL)
            {
                dmgRoles = new QVector<const Costume*>();
                actorData._dmg_roles = dmgRoles;
            }
            dmgRoles->append(&costume);
        }
        costume.refresh(ret, actor, false, false);
    }
}

void Costume::apply(QString& ret, Actor& actor) const
{
    const Costume& role = *this;
    ret += QString(Costume::CausesTxt).arg(actor.name(), role.name());
    role.damage(ret, NIL, actor, 0, true);
}

void Costume::refresh(QString* const ret, Actor& actor, bool const updStates, bool const remove) const
{
    const Costume& costume = *this;
    {
        auto costumeData = costume.costumeData();
        auto skills = costumeData._a_skills;
        if (skills)
        {
            actor.updateSkills(remove, false, *skills);
        }
        auto counters = costumeData._counters;
        if (counters)
        {
            actor.updateSkills(remove, true, *counters);
        }
    }
    if (updStates)
    {
        QList<Ailment>* cStates = costume.roleData()._a_states;
        if (cStates)
        {
            actor.updateStates(remove, ret, *cStates, false);
        }
    }
    if (!remove)
    {
        int const cFlags = costume.playFlags();
        if (costume.currentHp() == 0 && costume.currentMp() == 0 && costume.currentRp() == 0)
        {
            actor.setDamageType(actor.damageType() | costume.damageType());
        }
        actor.setPlayFlags(actor.playFlags() | cFlags);
        /*if (scene) //TODO: adapt without using "scene" pointer here
        {
#if ALLOW_COVERING
            if ((cFlags & Attribute::Covering) == Attribute::Covering)
            {
                scene->setHasCovers(true);
            }
#endif
            if (ret && (cFlags & Attribute::Ko) == Attribute::Ko)
            {
                scene->checkStatus(ret);
            }
        }*/
        if (costume.isShapeShifted())
        {
            QString* const spr = costume.roleData()._sprite;
            if (spr)
            {
                (*actor.roleData()._sprite) = *spr;
            }
        }
    }
}

Costume::CostumeSheet::~CostumeSheet() {}

Costume::CostumeSheet::CostumeSheet(int const id, QString& name, QString& sprite, bool const shapeShift, int const mActions, int const elm, int const hpDmg,
                                    int const mpDmg, int const spDmg, int const mHp, int const mMp, int const mSp, int const atk, int const def, int const spi,
                                    int const wis, int const agi, bool const stun, bool const range, bool const automate, bool const confuse, bool const reflect,
                                    bool const ko, bool const invincible, bool const revive, QList<Ability>* const skills, QList<Ability>* const counters,
                                    QList<Ailment>* const states, QMap<int, int>* const stRes, QMap<int, int>* const res)
    : RoleSheet(id, name, sprite, hpDmg, mpDmg, spDmg, mHp, mMp, mSp, elm, range, revive, states)
{
    this->_m_actions = mActions;
    this->_counters = counters;
    this->_a_skills = skills;
    this->_atk = atk;
    this->_def = def;
    this->_spi = spi;
    this->_wis = wis;
    this->_agi = agi;
    this->_res = res;
    this->_st_res = stRes;
    int flags = this->_play_flags;
    if (stun)
    {
        flags |= Attribute::STUN;
    }
    if (shapeShift)
    {
        QString* spr = this->_sprite;
        if (spr && spr->length() > 0)
        {
            flags |= Attribute::SHAPE_SHIFT;
        }
    }
    if (automate)
    {
        flags |= Attribute::ENRAGED;
    }
    if (confuse)
    {
        flags |= Attribute::CONFUSE;
    }
    if (reflect)
    {
        flags |= Attribute::REFLECT;
    }
    if (invincible)
    {
        flags |= Attribute::INVINCIBLE;
    }
    if (ko)
    {
        flags |= Attribute::KO;
    }
    this->_play_flags = flags;
}

Costume::Costume(int const id, QString& name, QString& sprite, bool const shapeShift, int const mActions, int const elm, int const hpDmg, int const mpDmg, int const spDmg, int const mHp,
                 int const mMp, int const mSp, int const atk, int const def, int const spi, int const wis, int const agi, bool const stun, bool const range, bool const automate, bool const confuse,
                 bool const reflect, bool const ko, bool const invincible, bool const revive, QList<Ability>* const skills, QList<Ability>* const counters, QList<Ailment>* const states,
                 QMap<int, int>* const stRes, QMap<int, int>* const res, QObject* const parent)
    : Costume(parent, new CostumeSheet(id, name, sprite, shapeShift, mActions, elm, hpDmg, mpDmg, spDmg, mHp, mMp, mSp, atk, def, spi, wis, agi, stun, range,
                                       automate, confuse, reflect, ko, invincible, revive, skills, counters, states, stRes, res)) {}

Costume::Costume(int const id, QString& name, QString& sprite, bool const shapeShift, int const mActions, int const elm, int const hpDmg, int const mpDmg, int const spDmg,
                 int const mHp, int const mMp, int const mSp, int const atk, int const def, int const spi, int const wis, int const agi, bool const range, bool const automate,
                 bool const confuse, bool const reflect, bool const invincible, bool const revive, QList<Ability>* const skills, QList<Ability>* const counters,
                 QList<Ailment>* const states, QMap<int, int>* const stRes, QMap<int, int>* const res, QObject* const parent)
    : Costume(id, name, sprite, shapeShift, mActions, elm, hpDmg, mpDmg, spDmg, mHp, mMp, mSp, atk, def, spi, wis, agi, false, range, automate, confuse, reflect, false, invincible,
              revive, skills, counters, states, stRes, res, parent) {}

Costume::Costume(int const id, QString&& name, QString&& sprite, bool const shapeShift, int const mActions, int const elm, int const hpDmg, int const mpDmg, int const spDmg,
                 int const mHp, int const mMp, int const mSp, int const atk, int const def, int const spi, int const wis, int const agi, bool const range, bool const automate,
                 bool const confuse, bool const reflect, bool const invincible, bool const revive, QList<Ability>* const skills, QList<Ability>* const counters,
                 QList<Ailment>* const states, QMap<int, int>* const stRes, QMap<int, int>* const res, QObject* const parent)
    : Costume(id, name, sprite, shapeShift, mActions, elm, hpDmg, mpDmg, spDmg, mHp, mMp, mSp, atk, def, spi, wis, agi, range, automate, confuse, reflect, invincible,
              revive, skills, counters, states, stRes, res, parent) {}

Costume::Costume(QObject* const parent) : Costume(0, QString(), QString(), false, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, false, false, false, false, false, false, NIL, NIL, NIL, NIL, NIL, parent) {}

Costume::Costume(QObject* const parent, CostumeSheet* const costumeDataPtr) : Role(parent, costumeDataPtr) {}

Costume::Costume(const Costume&& costume) : Costume(static_cast<const Costume&>(costume)) {}

Costume::Costume(const Costume& costume) : Role(costume) {}

Costume::~Costume() {}
